src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.cs

```diff
@@ -284,7 +284,29 @@ private void ChangeProjectProperty<T>(
 
                     if (!isFullyLoaded)
                     {
-                        TryReportCompilationThrownAway(_projectSystemProjectFactory.Workspace.CurrentSolution, Id);
+                        var currentSolution = _projectSystemProjectFactory.Workspace.CurrentSolution;
+                        var reportCompilationThrownAway = true;
+
+                        // Preprocessor directive only ParseOptions changes have special handling in DocumentState.UpdateParseOptionsAndSourceCodeKind.
+                        // We don't want to report telemetry for those changes or for the initial evaluation result.
+                        if (newValue is ParseOptions newParseOption)
+                        {
+                            if (oldValue is ParseOptions oldParseOptions)
+                            {
```

Comment:
Can we just do this logic in the ParseOptions setter and pass in logThrowAwayTelemetry appropriately? That way we're not doing this checks in other property tests, and we also skip all the stuff above which is probably not cheap either.

Reply:
I'm a bit hesitant to move it to the caller as the code is currently running inside the _gate lock. I'd like to understand the concern a bit better. The code is only executed when logThrowAwayTelemetry is true, so it's just two of the callers, and the added code is only executed before the solution finishes loading.

Reply:
Moving it out would also mean adding exception handling to ensure the field was still updated

Reply:
It's just a bit strange that this is applying to a specific case, but it's in the general helper.

> Moving it out would also mean adding exception handling to ensure the field was still updated

What do you mean?

Reply:
The issues I see if the logThrowAwayTelemetry block were called outside and prior to ChangeProjectProperty:

1) Exception handling would need to be added around the call to that block to ensure ChangeProjectProperty would continue to be called
2) The block would *probably* need to be executed under the gate's protection (at least it is now, and I haven't tried to suss out if it need be)

I've added some comments that make things a bit clearer, I'm definitely willing to change things if you think there is a better way.